### DISCUSSION TITLE
[Feature Request] add months in NumTimeExtension

### DISCUSSION THREAD

**Discussion by alfinbi** [jogboms/time.dart#55] [General]
please add months in NumTimeExtension

**Comment by shinriyo** [comment]
@alfinbi I also wanted years. but It is impossible.
Month's day's count are different depends month.
February is 28 or 29 days. And almost the month is 31 or 30 days.

That is why `1.months` are `28`, `29`, `30`, and`31`.

@jogboms Don't you think so?

**Comment by alfinbi** [comment]
I think we can use clampDayOfMonth method from package:quiver/time.dart and implement it as extension

**Comment by shinriyo** [comment]
@alfinbi however, `1.months` can't determine the number.

**Comment by alfinbi** [comment]
yeah I was just thinking about it, its hard to implement in duration extension, but I think it can be implemented in DateTime extension, for instance,

```
final now = DateTime.now();
final expiryDate  = now.addMonths(3);
final anotherDate = now.substractYears(1);
```

it will be useful for backend to calculate paid user expiry date. just like I have been done.

**Comment by shinriyo** [comment]
@alfinbi 
I added like this.
https://github.com/jogboms/time.dart/pull/26/commits/dd7e870fcb7c910271e010fa6ff2ce468aa32ca4

**Comment by jogboms** [comment]
Got swamped with daily work. Thinking about this, considering what @shinriyo already said, `1.month` is a bit meaningless as it it be either 28, 29, 30 or 31 which can in no way be determined automagically. I would rather this task be left on the developer level. I may be wrong.

**Comment by ChristianKleineidam** [comment]
It would be possible to create a new class for Month() and then allow addition for DateTime + Month.

**Comment by jogboms** [comment]
It ties in to overloading the `+` operator as you suggested earlier. In my head, I didn't want to introduce new types. 

`DateTime(2013, 1) + 6.months` -> `DateTime(2013, 1) + Month(6)`

Where `Month` can stay as a hidden detail

**Comment by sachaarbonel** [comment]
Hi guys I have a proposal and a test case.  Let's take the example of generating sales reports, for example, you need to iterate through months then weeks (4 in a month) then weekdays (can be 28,29, etc). I propose to introduce two enums (or more if you have other cases in mind) and their associated extensions:
```dart
main() {
      final dates = [
        DateTime(2020, 01, 31),
        DateTime(2020, 01, 31),
        DateTime(2020, 01, 31),
        DateTime(2020, 01, 31),
        DateTime(2020, 01, 31),
        DateTime(2020, 01, 31),
        DateTime(2020, 03, 15)
      ];

   final now = DateTime.now();
   for (var month in Month.values) {
        final sales = salesForMonth(dates, month.forYear(now.year));
        print("$sales sales in month ${month.toStr()}");
        print("details : ");
        for (var week in [1, 2, 3, 4]) {
          for (var weekDay in WeekDay.values) {
            final sales =
                salesForWeekDay(dates, weekDay.forMonth(month.forYear(now.year)));
            print(
                "$sales sales in week $week of month ${month.toStr()} day ${weekDay.toStr()}");
          }
        }
      }

}

int salesForWeekDay(List<DateTime> dates, DateTime day) =>
    (List<DateTime>.from(dates)..retainWhere((d) => d.sameWeekDay(day))).length;

int salesForMonth(List<DateTime> dates, DateTime month) =>
    (List<DateTime>.from(dates)..retainWhere((d) => d.sameMonth(month))).length;

extension DateHelpers on DateTime {
  bool sameYear(DateTime other) {
    return other.year == this.year;
  }

  bool sameMonth(DateTime other) {
    return sameYear(other) && other.month == this.month;
  }

  bool sameWeekDay(DateTime other) {
    return sameMonth(other) && other.weekday == this.weekday;
  }
}


enum Month {
  january,
  february,
  march,
  april,
  may,
  july,
  august,
  september,
  october,
  november,
  december,
}

extension MonthValue on Month {
  DateTime forYear(int year) {
    final y = DateTime(year);
    final month = DateTime(
        y.year,
        <Month, int>{
          Month.january: DateTime.january,
          Month.february: DateTime.february,
          Month.march: DateTime.march,
          Month.april: DateTime.april,
          Month.may: DateTime.may,
          Month.july: DateTime.july,
          Month.august: DateTime.august,
          Month.september: DateTime.september,
          Month.october: DateTime.october,
          Month.november: DateTime.november,
          Month.december: DateTime.december,
        }[this]);
    return month;
  }

  String toStr() => <Month, String>{
        Month.january: "january",
        Month.february: "february",
        Month.march: "march",
        Month.april: "april",
        Month.may: "may",
        Month.july: "july",
        Month.august: "august",
        Month.september: "september",
        Month.october: "october",
        Month.november: "november",
        Month.december: "december",
      }[this];
}


enum WeekDay {
  monday,
  tuesday,
  wednesday,
  thursday,
  friday,
  saturday,
  sunday,
}

extension DayValue on WeekDay {
  DateTime forMonth(DateTime month) {
    final day = DateTime(
        month.year,
        month.month,
        <WeekDay, int>{
          WeekDay.monday: DateTime.monday,
          WeekDay.tuesday: DateTime.tuesday,
          WeekDay.wednesday: DateTime.wednesday,
          WeekDay.thursday: DateTime.thursday,
          WeekDay.friday: DateTime.friday,
          WeekDay.saturday: DateTime.saturday,
          WeekDay.sunday: DateTime.sunday,
        }[this]);
    return day;
  }

  String toStr() => <WeekDay, String>{
        WeekDay.monday: "monday",
        WeekDay.tuesday: "tuesday",
        WeekDay.wednesday: "wednesday",
        WeekDay.thursday: "thursday",
        WeekDay.friday: "friday",
        WeekDay.saturday: "saturday",
        WeekDay.sunday: "sunday",
      }[this];
}

``` 
It should output this result : 
```bash
6 sales in month january
details : 
0 sales  in week 1  of month january day monday
0 sales  in week 1  of month january day tuesday
6 sales in week 1  of month january day wednesday

....
```
Some cons here are the toStr() which isn't ideal for translations but work great for debugging. And the naming conventions need some love
Tell me what you think @jogboms and I make the PR.

**Comment by tomtaila** [comment]
Would love this to be added please!

**Comment by shinriyo** [comment]
add, please.
or marge mine ;>

**Comment by jogboms** [comment]
I would pick this up again this weekend. I believe it aged well.

**Comment by tomtaila** [comment]
> add, please.
> or marge mine ;>

I use yours @shinriyo thank you very much for sharing it!

**Comment by jogboms** [comment]
Hey @sachaarbonel 
A lot of time has passed so I don't know if you are still interested on a quick PR. I am available for review. I feel its a new feature compared to this issue as #26 aims to solve this.

**Comment by FMorschel** [comment]
Maybe #50 and #51 can help?

**Comment by FMorschel** [comment]
With `lastDayOfMonth` you could potentially do something like
```dart
extension AddMonth on DateTime {
  DateTime addMonths([int amount = 1]) {
    final plusXMonths = DateTime(year, month + amount, day);
    final xMonthsAhead = DateTime(year, month + amount);
    if (xMonthsAhead.lastDayOfMonth.compareTo(plusXMonths).isNegative) {
      return xMonthsAhead.lastDayOfMonth;
    } else {
      return plusXMonths;
    }
  }
}
```

**Comment by FMorschel** [comment]
After some time and thinking, I don't believe this is the best approach anymore. I think this actually could be still confusing. 

Considering my PR #53 gets in, if not we could implement it here, maybe something like this instead would be better (maybe a better name but I couldn't think of one): 
```dart
static DateTime clampMonthLength(int year, int month, int day) {
  final firstDay = DateTime(year, month);
  return firstDay.copyWith(day: day).clamp(
        min: firstDay,
        max: firstDay.lastDayOfMonth,
      );
}
```

I also think there could be another method that would return a list of DateTimes with that day clamped to the end of the months, but that one is even harder for me to name and I'm not sure about if it would be a good fit and what it's signature would be.

**Reply by jogboms** [reply]
#53 is in now, great job @FMorschel

**Comment by FMorschel** [comment]
My final thoughts about this discussion (for now) are that this should not be inside the `DateTime` class.

I've created a new class named `DueDateTime` where we could save the _dueDay_ (int) and use this class to keep receiving the next months. I believe this extended class has a more specific purpose and would be better for future changes and understanding. 

<details>
  <summary>Click here to see my class.  I put it here under the <a href="https://opensource.org/licenses/MIT">MIT Licence</a>.</summary>
  
  ### Explanation for `copyWith`
  I'm always using `copyWith` so the `isUtc` property stays the same without having to test it every time.

  ### Code
```dart
class DueDateTime extends DateTime {
  factory DueDateTime(int dueDay, [DateTime? startReference]) {
    final reference = (startReference ?? DateTime.now()).date;
    late final DateTime dueDate;
    if (reference.day == dueDay) {
      return DueDateTime._toConstructor(dueDay, reference);
    } else if (reference.day < dueDay) {
      if (reference.lastDayOfMonth.day > dueDay) {
        dueDate = reference
            .to(reference.lastDayOfMonth)
            .firstWhere((date) => date.day == dueDay);
        return DueDateTime._toConstructor(dueDay, dueDate);
      } else {
        dueDate = reference
            .copyWith(day: dueDay)
            .clamp(max: reference.lastDayOfMonth);
        return DueDateTime._toConstructor(dueDay, dueDate);
      }
    } else {
      final dueNextMonth =
          reference.copyWith(month: reference.month + 1, day: dueDay);
      final endNextMonth =
          reference.copyWith(month: reference.month + 1, day: 1).lastDayOfMonth;
      dueDate = dueNextMonth.clamp(max: endNextMonth);
      return DueDateTime._toConstructor(dueDay, dueDate);
    }
  }

  DueDateTime._(this._dueDay, DateTime date)
      : super(
          date.year,
          date.month,
          date.day,
          date.hour,
          date.minute,
          date.second,
          date.millisecond,
          date.microsecond,
        );

  DueDateTime._utc(this._dueDay, DateTime date)
      : super.utc(
          date.year,
          date.month,
          date.day,
          date.hour,
          date.minute,
          date.second,
          date.millisecond,
          date.microsecond,
        );

  factory DueDateTime._toConstructor(int dueDay, DateTime date) {
    return date.isUtc
        ? DueDateTime._utc(dueDay, date)
        : DueDateTime._(dueDay, date);
  }

  int _dueDay;

  DueDateTime addMonths([int months = 1]) {
    final firstDay = copyWith(month: month + months, day: 1);
    return DueDateTime._toConstructor(
      _dueDay,
      firstDay.copyWith(day: _dueDay).clamp(
            min: firstDay,
            max: firstDay.lastDayOfMonth,
          ),
    );
  }
}
```
</details>

So wrapping up. Maybe my class would need to be changed for other use-cases but for mine, it was just what I needed. I don't believe this change should be added to this package unless we actually create a new class, forgetting the extensions for this feature.

**Reply by FMorschel** [reply]
@shinriyo, could you give us any feedback or thoughts? Maybe review [your PR](https://github.com/jogboms/time.dart/pull/33)?

**Comment by jogboms** [comment]
For anyone that arrives here. @FMorschel made a sweet package that covers all the cases mentioned here and more.
https://github.com/FMorschel/due_date
